public/nextjs_migration/client/js/circleChart.js

```diff
@@ -0,0 +1,326 @@
+}
+
+const html = () => `
+  ${styles}
+  <figure class='circle-chart'></figure>
+`
+
+customElements.define('circle-chart', class extends HTMLElement {
+  /** @type {Array<{ key: string; name: string; color: string; count: number; }> | null} */
+  data
+  /** @type {Element | null | undefined} */
+  chartElement
+  /** @type {string} */
+  showPercentFor
+  /** @type {string} */
+  title
+  /** @type {SVGSVGElement | null} */
+  svg
+
+  static get observedAttributes () {
+    return [
+      'data',
+      'show-percent-for',
+      'title'
+    ]
+  }
+
+  constructor () {
+    super()
+    this.attachShadow({ mode: 'open' })
+
+    // Chart properties
+    this.data = null
+    this.showPercentFor = ''
+    this.title = ''
+
+    this.svg = null
+    this.total = 0
+    this.updateTimeout = null
+
+    this.render()
+  }
+
+  /**
+   * @param {string} name
+   * @param {string} oldValue
+   * @param {string} newValue
+   */
+  attributeChangedCallback (name, oldValue, newValue) {
+    if (newValue === 'undefined' || newValue === oldValue) {
+      return
+    }
+
+    switch (name) {
+      case 'data':
+        this.data = JSON.parse(newValue)
+        this.createOrUpdateChart()
+        break
+      case 'show-percent-for':
+        this.showPercentFor = newValue
+        break
+      case 'title':
+        this.title = newValue
+        break
+      default:
+        console.warn(`Unhandled attribute: ${name}`)
+    }
+  }
+
+  composeCircles () {
+    let sliceOffset = 0
+    /** @type {string[]} */
+    const init = []
+    return `
+      ${this.data?.reduce((acc, curr) => {
+        const percentage = calcPercentage(this.total, curr.count)
+        const innerRadius = this.showPercentFor !== '' ? 0.85 : 0
+        const strokeLength = CHART_CIRCUMFERENCE * percentage
+
+        const circle = `
+          <circle
+            stroke='${curr.color}'
+            stroke-dasharray='${strokeLength} ${CHART_CIRCUMFERENCE}'
+            stroke-dashoffset='${-1 * CHART_CIRCUMFERENCE * sliceOffset}'
+            stroke-width='${CHART_DIAMETER * (1 - innerRadius)}%'
+          ></circle>
+        `
+        acc.push(circle)
+        sliceOffset += percentage
+
+        return acc
+      }, init).join('')}
+    `
+  }
+
+  createChartLabels () {
+    return `
+      ${this.title !== ''
+        ? `<strong class='circle-chart-title'>${this.title}</strong>`
+        : ''}
+      ${this.data?.map(({ name, color }) => (
+        `<label class='circle-chart-label' style='--color: ${color}'>${name}</label>`
+      )).join('')}
+    `
+  }
+
+  createCircleLabel () {
+    const relevantItem = this.data?.find(d => d.key === this.showPercentFor)
+    if (!relevantItem) {
+      return ''
+    }
+    const percentage = calcPercentage(this.total, relevantItem.count)
+    return `
+      <text
+        dy='${CHART_RADIUS * 0.15}'
+        fill='${relevantItem.color}'
+        font-size='${CHART_RADIUS * 0.4}'
+        font-size='50'
```

Comment:
`font-size` duped here... I'm guessing the first one (the calculated one) is probably ignored. 🤷

Reply:
Yupp — sneaked it’s way in: [b132b0c](https://github.com/mozilla/blurts-server/pull/3085/commits/b132b0c1cca01495aba0da23f82b2723572f0b21).